lib/Serialization/Deserialization.cpp

```diff
@@ -1469,8 +1469,87 @@ ModuleFile::resolveCrossReference(ModuleID MID, uint32_t pathLen) {
+        bool importedFromClang = false;
+        bool isStatic = false;
+        if (isType) {
+          XRefTypePathPieceLayout::readRecord(scratch, IID, privateDiscriminator,
+                                              inProtocolExt, importedFromClang);
+        } else {
+          XRefValuePathPieceLayout::readRecord(scratch, TID, IID, inProtocolExt,
+                                               importedFromClang, isStatic);
+        }
+
+        DeclBaseName name = getDeclBaseName(IID);
+        Type filterTy;
+        if (!isType) {
+          auto maybeType = getTypeChecked(TID);
+          // Any error here would have been handled previously.
+          if (maybeType) {
+            filterTy = maybeType.get();
+          }
+        }
+
+        values.clear();
+        if (privateDiscriminator) {
+          baseModule->lookupMember(values, baseModule, name,
+                                   getIdentifier(privateDiscriminator));
+        } else {
+          baseModule->lookupQualified(baseModule, DeclNameRef(name),
+                                      NL_QualifiedDefault,
+                                      values);
+        }
+
+        bool hadAMatchBeforeFiltering = !values.empty();
+        filterValues(filterTy, nullptr, nullptr, isType, inProtocolExt,
+                     importedFromClang, isStatic, None, values);
+
+        strScratch.clear();
+        auto declNameStr = declName.getString(strScratch).str();
+        if (!values.empty()) {
+          // Found a full match in a different module. It should be a different
+          // one because otherwise it would have succeeded on the first search.
+          // This is usually caused by the use of poorly modularized headers.
+          auto line = "'" + declNameStr + "' was not found in module '" +
+                      baseModule->getName().str() +
+                      "', but there is a '" +
+                      declNameStr + "' in module '" +
+                      nameAndModule.first.str() +
+                      "'. If this is imported from clang, please make sure " +
+                      "the header is part of a single clang module.";
+          notes.emplace_back(line.str());
+        } else if (hadAMatchBeforeFiltering) {
+          // Found a match that was filtered out. This may be from the same
+          // expected module if there's a type difference. This can be caused
+          // by the use of different Swift language versions between a library
+          // with serialized SIL and a client.
+          auto line = "'" + declNameStr + "' in module '" +
+                      baseModule->getName().str() +
+                      "' was filtered out.";
+          notes.emplace_back(line.str());
+        }
+      }
+    }
+
     return llvm::make_error<XRefError>("top-level value not found", pathTrace,
-                                       getXRefDeclNameForError());
+                                       declName, notes);
```

Comment:
Would it make sense to make this change to other XRefErrors too?

Reply:
The "top-level value not found" is the main error point that I've been seeing reporting such failures. I don't like so much the complexity this adds even currently so I'd prefer to make sure it's useful on other error points before adding more complexity.

The other XRefErrors above are more for Swift specific concepts like operators and precedence groups so modularization is more reliable with them. The errors below are for nested declarations so the search in other modules doesn't apply for the notes on filtering could be useful there.